vim_configurable build is broken on macOS

## Issue description

The `vim_configurable` derivation does not build on OS X 10.11.6:

```
builder for '/nix/store/sgqciz4jgsbvpliigh63gf9sh8varn5f-vim_configurable-8.1.0146.drv' failed with exit code 2; last 10 log lines:
                                 ^
  In file included from os_macosx.m:27:
  In file included from /nix/store/6w9yw7gqbfv5jd7r252k1r40wn3c9qsm-apple-framework-AppKit/Library/Frameworks/AppKit.framework/Headers/AppKit.h:203:
  /nix/store/6w9yw7gqbfv5jd7r252k1r40wn3c9qsm-apple-framework-AppKit/Library/Frameworks/AppKit.framework/Headers/NSPathComponentCell.h:42:27: warning: 'copy' attribute on property 'image' does not match the property inherited from 'NSCell' [-Wproperty-attribute-mismatch]
  @property (copy) NSImage *image;
                            ^
  6 warnings and 11 errors generated.
  make[1]: *** [Makefile:3260: objects/os_macosx.o] Error 1
  make[1]: Leaving directory '/private/tmp/nix-build-vim_configurable-8.1.0146.drv-0/source/src'
  make: *** [Makefile:29: first] Error 2
```

### Steps to reproduce

1. Update `nixkpgs` to `18.09pre149107.3878342ef02`
2. Setup a `darwin-configuration.nix` with `programs.vim.enable = true`
3. Call `darwin-rebuild switch`

## Technical details

```
$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-darwin"`
 - host os: `Darwin 15.6.0, macOS 10.11.6`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.0.4`
 - channels(JPC): `"darwin"`
 - channels(root): `"nixpkgs-18.09pre149107.3878342ef02"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixpkgs`
```

--- a/pkgs/applications/editors/vim/configurable.nix
+++ b/pkgs/applications/editors/vim/configurable.nix
@@ -24,9 +24,7 @@
 , cscopeSupport     ? config.vim.cscope or true     # Enable cscope interface
 , netbeansSupport   ? config.netbeans or true       # Enable NetBeans integration support.
 , ximSupport        ? config.vim.xim or true        # less than 15KB, needed for deadkeys
-# By default, compile with darwin support if we're compiling on darwin, but
-# allow this to be disabled by setting config.vim.darwin to false
-, darwinSupport     ? stdenv.isDarwin && (config.vim.darwin or true) # Enable Darwin support
+, darwinSupport     ? config.vim.darwin or false    # Enable Darwin support
 , ftNixSupport      ? config.vim.ftNix or true      # Add .nix filetype detection and minimal syntax highlighting support
 , ...
 }:
@@ -100,6 +98,8 @@ in stdenv.mkDerivation rec {
     "--disable-carbon_check"
     "--disable-gtktest"
   ]
+  ++ stdenv.lib.optional stdenv.isDarwin
+     (if darwinSupport then "--enable-darwin" else "--disable-darwin")
   ++ stdenv.lib.optionals luaSupport [
     "--with-lua-prefix=${lua}"
     "--enable-luainterp"
@@ -143,9 +143,6 @@ in stdenv.mkDerivation rec {
       cp ${vimPlugins.vim-nix.src}/indent/nix.vim runtime/indent/nix.vim
       cp ${vimPlugins.vim-nix.src}/syntax/nix.vim runtime/syntax/nix.vim
     '';
-
-  NIX_LDFLAGS = stdenv.lib.optionalString (darwinSupport && stdenv.isDarwin)
-    "/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation";
 
   postInstall = ''
   '' + stdenv.lib.optionalString stdenv.isLinux ''

--- a/pkgs/applications/editors/vim/configurable.nix
+++ b/pkgs/applications/editors/vim/configurable.nix
@@ -13,7 +13,7 @@
 
 , features          ? "huge" # One of tiny, small, normal, big or huge
 , wrapPythonDrv     ? false
-, guiSupport        ? config.vim.gui or "auto"
+, guiSupport        ? config.vim.gui or "gtk3"
 , luaSupport        ? config.vim.lua or true
 , perlSupport       ? config.vim.perl or false      # Perl interpreter
 , pythonSupport     ? config.vim.python or true     # Python interpreter

--- a/pkgs/applications/editors/vim/configurable.nix
+++ b/pkgs/applications/editors/vim/configurable.nix
@@ -1,7 +1,7 @@
 # TODO tidy up eg The patchelf code is patching gvim even if you don't build it..
 # but I have gvim with python support now :) - Marc
-args@{ source ? "default", callPackage, fetchurl, stdenv, ncurses, pkgconfig, gettext
-, writeText, lib, config, glib, gtk2, gtk3, python, perl, tcl, ruby
+{ source ? "default", callPackage, fetchurl, stdenv, ncurses, pkgconfig, gettext
+, writeText, config, glib, gtk2, gtk3, lua, python, perl, tcl, ruby
 , libX11, libXext, libSM, libXpm, libXt, libXaw, libXau, libXmu
 , libICE
 , vimPlugins
@@ -28,7 +28,8 @@ args@{ source ? "default", callPackage, fetchurl, stdenv, ncurses, pkgconfig, ge
 # allow this to be disabled by setting config.vim.darwin to false
 , darwinSupport     ? stdenv.isDarwin && (config.vim.darwin or true) # Enable Darwin support
 , ftNixSupport      ? config.vim.ftNix or true      # Add .nix filetype detection and minimal syntax highlighting support
-, ... }: with args;
+, ...
+}:
 
 
 let
@@ -100,7 +101,7 @@ in stdenv.mkDerivation rec {
     "--disable-gtktest"
   ]
   ++ stdenv.lib.optionals luaSupport [
-    "--with-lua-prefix=${args.lua}"
+    "--with-lua-prefix=${lua}"
     "--enable-luainterp"
   ]
   ++ stdenv.lib.optionals pythonSupport [
@@ -148,7 +149,7 @@ in stdenv.mkDerivation rec {
   postInstall = ''
   '' + stdenv.lib.optionalString stdenv.isLinux ''
     patchelf --set-rpath \
-      "$(patchelf --print-rpath $out/bin/vim):${lib.makeLibraryPath buildInputs}" \
+      "$(patchelf --print-rpath $out/bin/vim):${stdenv.lib.makeLibraryPath buildInputs}" \
       "$out"/bin/{vim,gvim}
 
     ln -sfn '${nixosRuntimepath}' "$out"/share/vim/vimrc

--- a/pkgs/applications/editors/vim/configurable.nix
+++ b/pkgs/applications/editors/vim/configurable.nix
@@ -129,7 +129,8 @@ in stdenv.mkDerivation rec {
 
   buildInputs = [ ncurses libX11 libXext libSM libXpm libXt libXaw libXau
     libXmu glib libICE ]
-    ++ (if guiSupport == "gtk3" then [gtk3] else [gtk2])
+    ++ stdenv.lib.optional (guiSupport == "gtk2") gtk2
+    ++ stdenv.lib.optional (guiSupport == "gtk3") gtk3
     ++ stdenv.lib.optionals darwinSupport [ CoreServices CoreData Cocoa Foundation libobjc cf-private ]
     ++ stdenv.lib.optional luaSupport lua
     ++ stdenv.lib.optional pythonSupport python

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -18836,7 +18836,6 @@ with pkgs;
   vim_configurable = vimUtils.makeCustomizable (callPackage ../applications/editors/vim/configurable.nix {
     inherit (darwin.apple_sdk.frameworks) CoreServices Cocoa Foundation CoreData;
     inherit (darwin) libobjc cf-private;
-    inherit lua;
     gtk2 = if stdenv.isDarwin then gtk2-x11 else gtk2;
     guiSupport = if stdenv.isDarwin then "gtk2" else "gtk3";
   });

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -9682,6 +9682,11 @@ with pkgs;
     inherit (darwin.apple_sdk.frameworks) AppKit Cocoa;
   };
 
+  # On darwin gtk uses cocoa by default instead of x11.
+  gtk3-x11 = gtk3.override {
+    x11Support = true;
+  };
+
   gtkmm2 = callPackage ../development/libraries/gtkmm/2.x.nix { };
   gtkmm3 = callPackage ../development/libraries/gtkmm/3.x.nix { };
 
@@ -18837,7 +18842,7 @@ with pkgs;
     inherit (darwin.apple_sdk.frameworks) CoreServices Cocoa Foundation CoreData;
     inherit (darwin) libobjc cf-private;
     gtk2 = if stdenv.isDarwin then gtk2-x11 else gtk2;
-    guiSupport = if stdenv.isDarwin then "gtk2" else "gtk3";
+    gtk3 = if stdenv.isDarwin then gtk3-x11 else gtk3;
   });
 
   qpdfview = libsForQt5.callPackage ../applications/misc/qpdfview {};

